Add initial version of paradox theme for akka docs

pvlugter opened the pull request:
Build akka docs using this branch:

https://github.com/akka/akka/tree/wip-akka-docs-new

patriknw commented:
@jrudolph @ktoso Google analytics tag is added here 3e1ccd9. We will get that for Akka Http also when we switch to this theme, I assume.

ktoso commented on theme/src/main/assets/analytics.st:
```diff
@@ -1,8 +1,8 @@
   var _gaq = _gaq || [];
-  _gaq.push(['_setAccount', 'UA-21117439-1']);
-  _gaq.push(['_setDomainName', 'akka.io']);
+  _gaq.push(['_setAccount', '$page.properties.("google.analytics.account")$']);
```
perfect, thanks @pvlugter

jonas commented on theme/src/main/assets/page.st:
```diff
@@ -0,0 +1,106 @@
+$page.content$
+</div>
+</article>
+
```
It would be nice to have the source URL template included here:

```html
               <div>
                 $source()$
               </div>
```

ktoso replied:
Ah indeed, that's an awesome feature!

pvlugter replied:
I'll add this.

jonas commented on theme/src/main/assets/page.st:
```diff
@@ -0,0 +1,106 @@
+<title>$page.title$$if(!page.home.active)$ &bull; $page.home.title$$endif$</title>
+<meta charset="utf-8"/>
+<meta name="viewport" content="width=device-width, initial-scale=1.0"/>
+<meta name="description" content="$page.properties.("project.description")$"/>
```
Would it make sense to apply the 'canonical' link support from https://github.com/akka/akka-http/pull/1091 here?

Completely untested code to exemplify:
```
$if(page.properties.("akka.canonical.base_url"))$
<link rel="canonical" href="$page.properties.("akka.canonical.base_url")$/$page.self.absolute.href$"/>
$endif$
```

pvlugter replied:
Yes.

jonas commented on theme/src/main/assets/page.st:
```diff
@@ -0,0 +1,106 @@
+<img class="akka-icon float-left show-for-medium" src="$page.base$images/akka-icon.svg">
+<section class="copyright">
+<div>&copy; 2011-$page.properties.("date.year")$ <a href="https://www.lightbend.com">Lightbend</a></div>
+<div>Akka is Open Source and available under the Apache 2 License.</div>
```
Maybe 

```
<div>$page.properties.("project.name")$ is Open Source and available under the Apache 2 License.</div>
```

to allow customization for Akka HTTP, Alpakka, etc.

pvlugter replied:
Good idea.

jonas commented:
Is the plan to add the Algolia doc search integration into the theme with per-project properties?

```sbt
paradoxProperties ++= Map(
  "algolia.docsearch.apiKey" -> "...",
  "algolia.docsearch.indexName" -> "..."
)
```

pvlugter commented:
@jonas, added the source code link, canonical link, project name in footer, and algolia search. For the search I combined what I saw in the akka and akka-http docs. Can you check these look good? 

Properties used for the akka docs for these:

```
name := "Akka"

paradoxProperties ++= Map(
  "akka.canonical.base_url" -> "http://doc.akka.io/docs/akka/current",
  "github.base_url" -> GitHub.url(version.value),
  "algolia.docsearch.api_key" -> "543bad5ad786495d9ccd445ed34ed082",
  "algolia.docsearch.index_name" -> "akka_io"
)
```

pvlugter commented:
I've added an sbt plugin as a way to have common paradox settings. This sets the theme and the navigation settings. And can be used to make sure that the theme is aligned with a particular version of paradox. So instead of using the sbt-paradox plugin directly and setting theme and navigation, use the sbt-paradox-akka plugin:

```
addSbtPlugin("com.lightbend.akka" % "sbt-paradox-akka" % "<version>")
```

```
enablePlugins(AkkaParadoxPlugin)
```

Currently uses snapshot version of paradox, and should be published.

patriknw commented:
@2m can you add this to our wip-akka-docs-new and publish needed snapshots?

2m commented:
Alright. Looking into it.

jonas commented:
@pvlugter Looks good. This covers the Paradox theme customizations in Akka HTTP.

jonas commented:
[Akka HTTP docs](https://jonas.github.io/akka-http/paradox-theme-akka/scala/http/index.html) compiled with the theme ([branch](https://github.com/jonas/akka-http/tree/paradox-theme-akka))

One question, should there be a way to differentiate between different Akka projects by looking at the docs, at least this is not clear when looking at the converted Akka docs. For Akka HTTP, we put the project name in the main index page so Paradox picks it up, but this is not perfect as can be seen below. Maybe this should be handled by this theme somehow. I think that would also be good for SEO.

<img width="1176" alt="screen shot 2017-05-12 at 4 44 20 pm" src="https://cloud.githubusercontent.com/assets/8417/26003407/e25f3ec8-3732-11e7-9a98-8f1a35866cc9.png">

ktoso commented:
Good point on how / where to differentiate hm.
Perhaps we can swap the left information of "Scala Documentation" with the "Akka" on the bottom, and make that project dependent - so top-left would always say "akka" / "akka http" / "alpakka" etc.

We have the font and can easily prepare "logos" for the other projects - so it would be in the right font too.

WDYT ?

pvlugter commented:
Yes, would be good to figure out the title / logo there. Could also do a simple rename in the Akka docs from "Scala Documentation" to "Akka (Scala)" or similar, so that it shows "Akka (Scala)" or "Akka (Java)". 

I also suggested an "akka docs" logo at the top at one point, which would go to the main akka docs page with links to the different modules.

pvlugter commented:
@raboof, added the group switching commits to this PR (and removed the `groups` branch). Also updated the style to fit both the group select and the search box.

pvlugter commented:
Added separate sidenote directive and styling. Note: this reduces the width of the content to have a margin on the right. Because of this, also made the width of the sidebar responsive so that it's not as wide when on the smaller end of 'large' screens. For not-large screens the sidenote is inlined.

For example, on full width screen:

![sidenote-large](https://cloud.githubusercontent.com/assets/59895/26044395/31da4878-3998-11e7-9e3f-40ced1dd0d05.png)

On smaller screen but still within the large screen min-width (with sidebar width reduced):

![sidenote-just-large](https://cloud.githubusercontent.com/assets/59895/26044403/376749bc-3998-11e7-97de-0c02b2b80966.png)

On small screen:

![sidenote-small](https://cloud.githubusercontent.com/assets/59895/26044470/a24f8dc0-3998-11e7-92fb-a18f0a2dd55b.png)

pvlugter commented:
Sidenote directive requires this version of paradox: https://github.com/lightbend/paradox/pull/115

ktoso commented:
> Yes, would be good to figure out the title / logo there. Could also do a simple rename in the Akka docs from "Scala Documentation" to "Akka (Scala)" or similar, so that it shows "Akka (Scala)" or "Akka (Java)".

Yeah that'd be a simple first step - sounds good :)

jonas commented:
I just noticed that the index overlay is slightly off on smaller screens
![screenshot_20170515-215522](https://cloud.githubusercontent.com/assets/8417/26076720/f9127066-39b9-11e7-92e2-1b0118caac15.png)

pvlugter commented:
@jonas, I don't see that extra offset for the overlay. I've made some updates to the overlay and added the navigation header there. Possibly simpler and hopefully better browser support. Could you try this in the same browser:

https://downloads.lightbend.com/paradox/akka-docs/20170516-nav-overlay/scala/actors.html

I don't work with CSS often, so if anyone has a good approach for doing this...

pvlugter commented:
Paradox 0.2.11 is published with the changes that have been used here. Snapshot of paradox is no longer needed.

jonas commented:
@pvlugter That looks much better.

pvlugter commented:
@jonas cool, thanks for checking.

pvlugter commented:
Should we require a class on the TOC to apply the block grid styling? It's currently applied to any TOC in a page, but there are places where it could be better as a regular list. For example:

https://jonas.github.io/akka-http/paradox-theme-akka/scala/http/routing-dsl/index.html

Or we can do it the other way around, and have a class that disables the box style on a TOC.

patriknw commented:
it would be useful to be able to choose between the two styles, perhaps keep the block style as default?

jonas commented:
Having the ability to choose the style with the `class` attribute would be great. I also wonder if the list style should be applied by default when the depth is 1 although that is probably lost in the translation to HTML. Akka HTTP has a lot of ToCs with depth 1 as part of the directives doc.

jonas commented on theme/src/main/assets/page.st:
```diff
@@ -0,0 +1,127 @@
+<link rel="stylesheet" type="text/css" href="$page.base$lib/foundation/dist/css/foundation.min.css"/>
+<link rel="stylesheet" type="text/css" href="$page.base$css/icons.css"/>
+<link rel="stylesheet" type="text/css" href="$page.base$css/page.css"/>
+<link rel="shortcut icon" href="$page.base$images/favicon.ico" />
```
Mobile browser tweaks as per https://developers.google.com/web/fundamentals/design-and-ui/browser-customization/
```
<meta name="theme-color" content="#15a9ce">
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
```

pvlugter replied:
Sounds good. I'll add these.

pvlugter commented:
@jonas, I like the idea of defaulting to list style for TOCs with depth 1.

It could also be useful for some paradox directives to set data attributes in the HTML. The general-purpose directives like `div` could pass through `data-*` attributes along with the class attribute.

pvlugter commented:
Updated the TOC styling. The default is block grid for TOCs with depth > 1 and regular list style for depth 1. Can force the styling to be block grid by adding `.blocks` class attribute or list style with `.list` class.

pvlugter commented:
Tweaked the TOC styling some more. There are now three supported styles:

* blocks: as now, default for TOCs with depth > 1
* box: a single full-width box for the TOC, default for depth 1
* list: a regular list of links

Can force to a particular style by adding class attribute. Also distinguish page and header links, like in the navigation.

pvlugter commented:
Added some styling for the search (which wasn't displaying the autocomplete results) and added search to the small screen navigation.

bantonsson commented:
Since we seem to be tweaking this a lot, is it only me or is the code block font size/spacing way bigger than the normal text?

For me this looks more in line with the text size:
```
code {
  font-size: 0.9em !important;
  line-height: 0.95em !important;
}
```

pvlugter commented:
Just when I was going to declare the tweaking done 😉 We haven't changed the font sizes here from what the designer put together. Source Code Pro certainly looks bigger to me too, even though the body and code font sizes are both at 1em. In the Cinnamon docs we have the code font size smaller for this reason too. Not sure on the conventions, but I guess it makes since to go 0.9375em to take a default 16px down to 15px. I'll update.

pvlugter commented:
Okay, I think this is ready to be the first version of the theme now. If it's looking good, then can be merged and first final release published.

raboof commented:
If we can also merge https://github.com/pvlugter/akka-paradox/pull/2 then I think we're good to go :)
